### PR TITLE
Use std::vector instead of Table

### LINKED ISSUE #12993: Memory leak with simplex meshes
Hi all!
@michelebucelli and I discovered a possible memory leak arising when using simplex meshes, as already reported to @peterrum.

# MWE
We ended up with the following MWE: [memory_leak.zip](https://github.com/dealii/dealii/files/7602940/memory_leak.zip)

The code just instantiates a FE vector and calls `DataOut::build_patches()` for, _e.g._, 1000 times.

The preprocessor variable `WITH_SIMPLEX` can be set to `1` or `0` to run the code on a simplex or hexahedral mesh, respectively.
In the former case, the memory consumption keeps increasing with time, whereas the latter seems to work like a charm.

# Results
Please find attached the source code, the corresponding `CMakeLists.txt` and a sample script to run `valgrind` using the heap profiler `massif`. Two graphical outputs are attached below.

After inspecting the `massif` output, our initial **guess** is that some memory is lost inside the `MappingFE::clone()` method whenever `FE_SimplexP` is used instead of `FE_Q`.

## Simplex mesh
memory_usage_simplex
## Hexahedral mesh
memory_usage_hex

### PR DESCRIPTION
**Pull request by elauksap**
Follow-up from #12993.

This fix bypasses the memory leak, which by the way still requires further investigation.

closes #12993

### CODE REVIEW

File: include/deal.II/base/polynomials_barycentric.h [hunk 1]
```diff
@@ -215,6 +215,31 @@ template <int dim>
...
+
+  /**
+   * Alias for polynomial third derivatives type.
+   */
+  using ThirdDerivativesType = std::array<HessianType, dim>;
+
+  /**
+   * Alias for polynomial fourth derivatives type.
+   */
+  using FourthDerivativesType = std::vector<ThirdDerivativesType>;
```
Comment (kronbichler) [inline include/deal.II/base/polynomials_barycentric.h:241]: Did you mean to use
```suggestion
  using FourthDerivativesType = std::array<ThirdDerivativesType, dim>;
```
here as well?

**Review by kronbichler — APPROVED** [review]
I think this is a good local solution. Next we need to find out the problem for the `AlignedVector`.

File: include/deal.II/base/polynomials_barycentric.h (hunk 1 above)
Comment (elauksap) [reply include/deal.II/base/polynomials_barycentric.h]: Indeed, it was just an oversight! Thanks for spotting it out :+1:

**Comment by peterrum** [comment]
@elauksap What is the reason that this PR is still "draft"?

**Comment by elauksap** [comment]
> @elauksap What is the reason that this PR is still "draft"?

Just removed that label :wink: I was waiting for the tests on my local machine to complete.

**Review by peterrum — APPROVED** [review]

**Comment by bangerth** [comment]
I like the solution with the statically typed array. Would this be something we could also apply to other similar classes? If so, want to write a patch?

**Comment by bangerth** [comment]
I was specifically thinking of the other polynomial classes.

**Comment by elauksap** [comment]
> I was specifically thinking of the other polynomial classes.

Ok, I see! You mean to replace, _e.g._, `Tensor<1, dim>` by `std::array<GradType, dim>` and so on like [here](https://github.com/dealii/dealii/blob/24fafe1087bf62fe27d73dac9e7b3b81a195c141/include/deal.II/base/scalar_polynomials_base.h#L102) and in all similar places, right?

**Comment by bangerth** [comment]
I think that doesn't work because you don't know how many basis functions you have, and so it really needs to be a dynamic `std::vector` or `Table` rather than a fixed size `std::array`. The beauty of your patch is that it avoids the dynamic allocation of memory by a statically sized object.

I don't know whether there are other similar opportunities in the classes I mentioned. It's possible that there are none, I was just wondering whether there are :-)

**Comment by elauksap** [comment]
As far as I can see, after some searching and `grep`ping, there is no other polynomial class where this patch can be applied.

**Comment by bangerth** [comment]
OK, thanks for checking!

**Pull request merged**